esp-hal/src/lock.rs

```diff
@@ -210,6 +212,31 @@ pub(crate) fn lock<T>(lock: &Lock, f: impl FnOnce() -> T) -> T {
     f()
 }
 
+/// Data protected by [LockState]
+#[allow(unused)]
+pub(crate) struct Locked<T> {
+    lock_state: Lock,
+    data: UnsafeCell<T>,
+}
+
+#[allow(unused)]
+impl<T> Locked<T> {
+    /// Create a new instance
+    pub(crate) const fn new(data: T) -> Self {
+        Self {
+            lock_state: Lock::new(),
+            data: UnsafeCell::new(data),
+        }
+    }
+
+    /// Provide exclusive access to the protected data to the given closure
+    pub(crate) fn with<R>(&self, f: impl FnOnce(&mut T) -> R) -> R {
+        lock(&self.lock_state, || f(unsafe { &mut *self.data.get() }))
+    }
```

Comment:
My bad, this is unfortunately unsound as `Lock` is re-entrant on single core chips. :sweat_smile:

It just needs an `UnsafeCell<bool>` to keep track of whether it's locked or not.

Reply:
We discussed this here btw: https://github.com/esp-rs/esp-hal/pull/2178#discussion_r1766733677, panicking in the case of reentrancy is reasonable enough for a internal API.